api/mymonero_rest.md

```diff
@@ -0,0 +1,391 @@
+| start_height         | uint64                 | Start height of response  |
+| blockchain_height    | uint64                 | Current blockchain height |
+| transactions         | array of transaction's | Possible spend info       |
+
+#### `get_random_outs`
+Selects random outputs to use in a ring signature of a new transaction. If the
+`amount` is `0` then the `monerod` RPC `get_output_distribution` should be used
+to locally select outputs using a gamma distribution as described in "An
+Empirical Analysis of Traceability in the Monero Blockchain". If the `amount`
+is not `0`, then the `monerod` RPC `get_output_histogram` should be used to
+locally select outputs using a triangular distribution
+(`uint64_t dummy_out = histogram.total * sqrt(float64(random_uint53) / float64(2^53))`).
+
+**Request** object
+
+|    Field   |            Type           |          Description             |
+|------------|---------------------------|----------------------------------|
+| count      | uint32                    | Mixin (name is historical)       |
+| amounts    | array of uint64-strings's | XMR amounts that need mixing     |
+
+> Clients must use amount `0` when computing a ringct output.
+
+> If clients are creating multiple rings with the same amount, they must set
+> `count` to the mixin level and add the value to `amounts` multiple times.
+> Server must respond to each value in `amounts`, even if the value appears
+> multiple times.gi
+
+**Response** object
+
+|    Field    |           Type           |          Description             |
+|-------------|--------------------------|----------------------------------|
+| amount_outs | array of random_outputs' | Dummy outputs for each `amounts` |
+
+> If there are not enough outputs to mix for a specific amount, the server
+> shall omit the `outputs` field in `amount_outs`.
+
+#### `get_unspent_outs`
+Returns a list of received outputs. The client must determine when the output
+was actually spent.
+
+**Request** object
+
+|       Field      |      Type      |           Description            |
+|------------------|----------------|----------------------------------|
+| address          | base58-address | Address to create/probe          |
+| view_key         | binary         | View key bytes                   |
+| amount           | uint64-string  | XMR send amount                  |
+| mixin            | uint32         | Minimum mixin for source output  |
+| use_dust         | boolean        | Return all available outputs     |
+| dust_threshold * | uint64-string  | Ignore outputs below this amount |
+
+> If the total received outputs for the address is less than `amount`, the
+> server shall return a HTTP 400 "Bad Request" error code.
+
+**Response** object
+
+|    Field   |        Type       |                Description              |
+|------------|-------------------|-----------------------------------------|
+| per_kb_fee | uint64            | Estimated network fee                   |
```

Comment:
Hi. It seems that we'll want to deprecate this field in favor of a `per_byte_fee`, right?

If so, we've unfortunately already released a client using the existing field, which we don't want to break. So we'd like to do this in a backward compatible way. 

Since ideally (IMO) we only want to send this old field to old clients, I would propose adding an optional field to this `get_unspent_outs` Request object for new clients to identify themselves. This could just be a feature flag, but I think it'd probably be better to just introduce an API version number here. In that case, the API exactly as documented here would constitute version 1, and a proposed version 2 would replace the Response field.

What are everyone's opinions about this?

Reply:
> Hi. It seems that we'll want to deprecate this field in favor of a per_byte_fee, right?

I recently proposed an update to the ZMQ call for fees that returns `estimated_base_fee`, `size_scale`, `fee_mask`, and `hard_fork_version`. The size scale returns `1024` when per-kb and `1` when per-byte. The fee mask is fixed for all versions, and seems unlikely to change anytime soon (ha!).

The per-byte is sufficient, but the client would have to know the fee mask locally or some fees will be slightly too low. Sending the `hard_fork_version` might be useful in future forks.

> p.s. why per_kb_fee is not string type? Shouldn't all xmr amounts uint64_t be represented as strings?

I'm not certain why this is a `uint64` field (before my time), but the `per_kb_fee` is unlikely to be larger than 2^53. I'm not opposed to changing it for consistency purposes if its somehow easier on someone.

Reply:
> Since ideally (IMO) we only want to send this old field to old clients, I would propose adding an optional field to this get_unspent_outs Request object for new clients to identify themselves. This could just be a feature flag, but I think it'd probably be better to just introduce an API version number here. In that case, the API exactly as documented here would constitute version 1, and a proposed version 2 would replace the Response field.

The server could continue to send `per_kb_fee` for backwards compatibility. I can think of a reason to immediately remove this field.

Reply:
Actually, just disregard the backwards-compatibility stuff. We've decided we don't want to support the old client after all (it has an unrelated issue when sending transactions). So, we can simply replace the field now.

Reply:
Unless I'm mistaken, we also need to send the fee quantization mask to the client, right?

In wallet2, this mask is retrieved from the daemon ([here](https://github.com/monero-project/monero/blob/8534f71eedbb421ca3e961c5adc96e5048c4e685/src/wallet/wallet2.cpp#L8404)) and applied ([here](https://github.com/monero-project/monero/blob/8534f71eedbb421ca3e961c5adc96e5048c4e685/src/wallet/wallet2.cpp#L236) via [here](https://github.com/monero-project/monero/blob/8534f71eedbb421ca3e961c5adc96e5048c4e685/src/wallet/wallet2.cpp#L8718)) after multiplying by the final tx size.